AccessControl: Document basic roles changes and provisioning V2 (#48910)
grafana/grafana: docs/sources/enterprise/access-control/manage-rbac-roles.md

This section includes instructions for how to view permissions associated with roles, create custom roles, and update and delete roles.

**Before you begin:**

- [Enable role-based access control]({{< relref "./enable-rbac-and-provisioning#enable-rback" >}}).

The following example includes the base64 username:password Basic Authorization. You cannot use authorization tokens in the request.

### List permissions associated with roles

Use a `GET` command to see the actions and scopes associated with a role. For more information about seeing a list of permissions for each role, refer to [Get a role]({{< relref "../../http_api/access_control.md#get-a-role" >}}).

> If you want to see the permissions associated to basic roles, their uid follows this mapping:

Reviewer: ```suggestion
To see the permissions associated with basic roles, refer to the following basic role UIDs:
```

| `global`      | Global roles are not associated with any specific organization, which means that you can reuse them across all organizations. This setting overrides `orgId`.                                                                                                                                                                                                                                                                                      |
| `displayName` | Human-friendly text that is displayed in the UI. Role display name cannot be longer than 190 ASCII-based characters. For fixed roles, the display name is shown as specified. If you do not set a display name the display name replaces `':'` (a colon) with `' '` (a space).                                                                                                                                                                     |

> | `description` | A human-friendly text to describe the permissions a role provides.                                                                                                                                                                                                                                                                                                                                                                                 |

Reviewer: ```suggestion
| `description` | Human-friendly text that describes the permissions a role provides.                                                                                                                                                                                                                                                                                                                                                                                 |
```

| `state`       | State of the role. Defaults to `present`, but if set to `absent` the role will be removed.                                                                                                                                                                                                                                                                                                                                                         |
| `force`       | Can be used in addition to state `absent`, to force the removal of a role and all its assignments.                                                                                                                                                                                                                                                                                                                                                 |

> | `from`        | Optional list of roles we want to copy permissions from.                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer: ```suggestion
| `from`        | An optional list of roles from which you want to copy permissions.                                                                                                                                                                                                                                                                                                                                                                                           |
```

```yaml
# config file version
apiVersion: 2

roles:
  - name: custom:users:writer
    description: 'List, create, or update other users.'
    version: 1
    orgId: 1
    permissions:
      - action: 'users:read'
        scope: 'global.users:*'
      - action: 'users:write'
        scope: 'global.users:*'
      - action: 'users:create'
```

The following example creates a hidden global role. The `global: true` option creates a global role, and the `hidden: true` option hides the role from the role picker.

```yaml
# config file version
apiVersion: 2

roles:
  - name: custom:users:writer
    description: 'List, create, or update other users.'
    version: 1
    global: true
    hidden: true
    permissions:
      - action: 'users:read'
        scope: 'global.users:*'
      - action: 'users:write'
        scope: 'global.users:*'
      - action: 'users:create'
```

> The following example creates a global role based on other fixed roles. The `from` option contains the roles we want to

Reviewer: ```suggestion
The following example creates a global role based on other fixed roles. The `from` option contains the roles from which we want to
```

> copy permissions from. The permission `state: absent` option allows to specify permissions to exclude from the copy.

Reviewer: ```suggestion
copy permissions. The permission `state: absent` option can be used to specify permissions to exclude from the copy.
```

Refer to the [RBAC HTTP API]({{< relref "../../http_api/access_control.md#create-a-new-custom-role" >}}) for more details.

> ## Update a basic role permissions

Reviewer: ```suggestion
## Update basic role permissions
```

**Before you begin:**

<!-- TODO change the link once we change assignments -->

- Determine the permissions you want to add or remove from a basic role. For more information about the permissions associated with basic roles, refer to [RBAC role definitions]({{< relref "./rbac-fixed-basic-role-definitions#basic-role-assignments" >}}).

</br>

**To change permissions from a basic role:**

1. Open the YAML configuration file and locate the `roles` section.

1. Refer to the following table to add attributes and values.

   | Attribute             | Description                                                                                                                               |
   | --------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
   | `name`                | The name of the basic role you want to update. You can specify a `uid` instead of a role name. The role `name` or the `uid` are required. |
   | `orgId`               | Identifies the organization to which the role belongs. `global` can be used instead to specify it's a global role.                        |

>    | `version`             | Identifies version of the role, to prevent overwriting newer changes.                                                                     |

Reviewer: ```suggestion
   | `version`             | Identifies the version of the role, which prevents overwriting newer changes.                                                                     |
```

>    | `from`                | List of roles to copy permissions from.                                                                                                   |

Reviewer: ```suggestion
   | `from`                | List of roles from which to copy permissions.                                                                                                   |
```

   | `permissions > state` | The state of the permission. You can set it to `absent` to ensure its exclusion from the copy list.                                       |

1. Reload the provisioning configuration file.

   For more information about reloading the provisioning configuration at runtime, refer to [Reload provisioning configurations]({{< relref "../../http_api/admin/#reload-provisioning-configurations" >}}).

The following example modifies the `Grafana Admin` basic role permissions.

> It removes from it, the permission to list, grant and revoke roles to teams.

Reviewer: ```suggestion
- Permission to list, grant, and revoke roles to teams is removed.
```

> It also adds to it the permission to read and write grafana folders.

Reviewer: ```suggestion
Permission to read and write Grafana folders is added.
```

```yaml
# config file version
apiVersion: 2

roles:
  - name: 'basic:grafana_admin'
    global: true
    version: 3
    from:
      - name: 'basic:grafana_admin'
        global: true
    permissions:
      # Permissions to remove
      - action: 'teams.roles:list'
        scope: 'teams:*'
        state: 'absent'
      - action: 'teams.roles:remove'
        scope: 'permissions:type:delegate'
        state: 'absent'
      - action: 'teams.roles:add'
        scope: 'permissions:type:delegate'
        state: 'absent'
      # Permissions to add
      - action: 'folders:read'
        scope: 'folder:*'
      - action: 'folders:write'
        scope: 'folder:*'
```

> **Note**: You can add multiple `fixed`, `basic` or `custom` roles to the `from` section. Their permissions will be copied and added to the basic role.
> <br/> > **Note**: Make sure to **increment** the role version for the changes to be accounted for.

You can also change basic roles' permissions using the API. Refer to the [RBAC HTTP API]({{< relref "../../http_api/access_control.md#update-a-role" >}}) for more details.

> ## Reset basic roles permissions to factory

Author: ```suggestion
## Reset basic roles permissions
```

> ## Reset basic roles permissions to factory

Reviewer: I've seen 'factory' elsewhere and don't think we should use that word. Our software was not designed in a factory. Can we simply say 'default settings' or something like that?

> This section describes how to reset the basic roles permissions to factory defaults:

Reviewer: same comment here

Author: ```suggestion
This section describes how to reset the basic roles permissions to defaults:
```

1. Open the YAML configuration file and locate the `roles` section.

1. Grant the `action: "roles:write", scope: "permissions:type:escalate` permission to `Grafana Admin`.

   ```yaml
   apiVersion: 2
   roles:
     - name: 'basic:grafana_admin'
       global: true
       version: 3
       from:
         - name: 'basic:grafana_admin'
           global: true
       permissions:
         # Permission allowing to reset basic roles to factory
         - action: 'roles:write'
          scope: 'permissions:type:escalate'
   ```

>    **Note**: This permission has not been granted to any basic roles by default, since it could allow in some cases, privilege escalations.

Reviewer: ```suggestion
> **Note**: This permission has not been granted to any basic roles by default, because in some cases it does not allow privilege escalations.
```

Author: I think this changes what I'm trying to express.Would this be better:

```suggestion
   **Note**: This permission has not been granted to any basic roles by default, since users could acquire more permissions than they previously had through the basic role permission reset.
```

> 1. As a `Grafana Admin`, call the API endpoint to reset the basic roles to factory. Refer to the [RBAC HTTP API]({{< relref "../../http_api/access_control.md#reset-basic-roles-permissions-to-factory" >}}) for more details.

Reviewer: Same comment about 'factory'.
